### PR TITLE
Add a user_data provisioning type (http://theforeman.org/issues/1561)

### PR DESCRIPTION
**Pull request by pccowboy**
This has been lightly tested, with a provisioning template of 

```
#cloud-config
repo_releasever: 2011.09
```

I will testing it more on my development puppet box over the next week, I am hoping you find it useful.

Not sure why there are 8 commits, though - you already merged those.  Sorry for the trouble.

### CODE REVIEW

**Comment by ohadlevy** [comment]
how do we handle certs this way?

**Comment by pccowboy** [comment]
I am currently using auto-sign, but cloud-config has a way to send the 
cert, detailed at 
http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/view/head:/doc/examples/cloud-config-ca-certs.txt

#cloud-config
#

# This is an example file to configure an instance's trusted CA certificates

# system-wide for SSL/TLS trust establishment when the instance boots

for the

# first time.

#

# Make sure that this file is valid yaml before starting instances.

# It should be passed as user-data when starting the instance.

ca-certs:
   # If present and set to True, the 'remove-defaults' parameter will remove
   # all the default trusted CA certificates that are normally shipped with
   # Ubuntu.
   # This is mainly for paranoid admins - most users will not need this
   # functionality.
   remove-defaults: true

   # If present, the 'trusted' parameter should contain a certificate 
(or list
   # of certificates) to add to the system as trusted CA certificates.
   # Pay close attention to the YAML multiline list syntax.  The example 
shown
   # here is for a list of multiline certificates.
   trusted:
- |
  -----BEGIN CERTIFICATE-----
  YOUR-ORGS-TRUSTED-CA-CERT-HERE
  -----END CERTIFICATE-----
- |
  -----BEGIN CERTIFICATE-----
  YOUR-ORGS-TRUSTED-CA-CERT-HERE
  -----END CERTIFICATE-----

Ohad Levy wrote:

**Comment by pccowboy** [comment]
I found what I think is a bug in a commit from abenari - commit https://github.com/pccowboy/foreman/commit/5051c7a00bc31325825e62794157c6a3183c32f6 has a (poor) fix.

File: app/models/orchestration/ssh_provision.rb [hunk 1]
```diff
@@ -92,7 +92,8 @@ def validate_ssh_provisioning
     return if Rails.env == "test"
     status = true
     begin
-      configTemplate(:kind => "finish")
+      template = configTemplate(:kind => "finish")
+      status = (template != nil)
```
Comment (ohadlevy) [inline app/models/orchestration/ssh_provision.rb:104]: yep, however I cant merge your current pull request, either you create another one, or I'll cherry pick the commit from here.

**Comment by ohadlevy** [comment]
i meant puppet ca, but, do you think its considered a trusted source to put the  certificate private key in exiternal userdata / metadata service?

**Comment by pccowboy** [comment]
In the case of EC2, I think so, since all communication to AWS is encrypted, and the userdata can only get to your instance if you have the correct AWS security credentials.  I will ask their support staff as well, just to make sure that the userdata is not logged anywhere else at all.

**Comment by pccowboy** [comment]
I am going to close this request, and open it after a vigorous rebase.

**Pull request closed without merging**